pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1633 @@
+		})
+	} else {
+		tx.Flush(&knftables.Set{
+			Name: kubeNodePortIPsSet,
+		})
+		nodeIPs, err := proxier.nodePortAddresses.GetNodeIPs(proxier.networkInterfacer)
+		if err != nil {
+			klog.ErrorS(err, "Failed to get node ip address matching nodeport cidrs, services with nodeport may not work as intended", "CIDRs", proxier.nodePortAddresses)
+		}
+		for _, ip := range nodeIPs {
+			if ip.IsLoopback() {
+				klog.ErrorS(nil, "--nodeport-addresses includes localhost but localhost NodePorts are not supported", "address", ip.String())
+				continue
+			}
+			tx.Add(&knftables.Element{
+				Set: kubeNodePortIPsSet,
+				Key: []string{
+					ip.String(),
+				},
+			})
+		}
+	}
+
+	// Set up "no endpoints" drop/reject handling
+	tx.Add(&knftables.Map{
+		Name:    kubeNoEndpointServicesMap,
+		Type:    ipvX_addr + " . inet_proto . inet_service : verdict",
+		Comment: ptr.To("vmap to drop or reject packets to services with no endpoints"),
+	})
+	tx.Add(&knftables.Map{
+		Name:    kubeNoEndpointNodePortsMap,
+		Type:    "inet_proto . inet_service : verdict",
+		Comment: ptr.To("vmap to drop or reject packets to service nodeports with no endpoints"),
+	})
+
+	tx.Add(&knftables.Chain{
+		Name:    kubeRejectChain,
+		Comment: ptr.To("helper for @no-endpoint-services / @no-endpoint-nodeports"),
+	})
+	tx.Flush(&knftables.Chain{
+		Name: kubeRejectChain,
+	})
+	tx.Add(&knftables.Rule{
+		Chain: kubeRejectChain,
+		Rule:  "reject",
+	})
+
+	tx.Add(&knftables.Rule{
+		Chain: kubeEndpointsCheckChain,
+		Rule: knftables.Concat(
+			ipX, "daddr", ".", "meta l4proto", ".", "th dport",
+			"vmap", "@", kubeNoEndpointServicesMap,
+		),
+	})
+
+	if proxier.nodePortAddresses.MatchAll() {
+		tx.Add(&knftables.Rule{
+			Chain: kubeEndpointsCheckChain,
+			Rule: knftables.Concat(
+				"fib daddr type local",
+				noLocalhost,
+				"meta l4proto . th dport",
+				"vmap", "@", kubeNoEndpointNodePortsMap,
+			),
+		})
+	} else {
+		tx.Add(&knftables.Rule{
+			Chain: kubeEndpointsCheckChain,
+			Rule: knftables.Concat(
+				ipX, "daddr", "@", kubeNodePortIPsSet,
+				"meta l4proto . th dport",
+				"vmap", "@", kubeNoEndpointNodePortsMap,
+			),
+		})
+	}
+
+	// Set up LoadBalancerSourceRanges firewalling
+	tx.Add(&knftables.Set{
+		Name:    kubeFirewallSet,
+		Type:    ipvX_addr + " . inet_proto . inet_service",
+		Comment: ptr.To("destinations that are subject to LoadBalancerSourceRanges"),
+	})
+	tx.Add(&knftables.Set{
+		Name:    kubeFirewallAllowSet,
+		Type:    ipvX_addr + " . inet_proto . inet_service . " + ipvX_addr,
+		Flags:   []knftables.SetFlag{knftables.IntervalFlag},
+		Comment: ptr.To("destinations+sources that are allowed by LoadBalancerSourceRanges"),
+	})
+
+	ensureChain(kubeFirewallCheckChain, tx, createdChains)
+	ensureChain(kubeFirewallAllowCheckChain, tx, createdChains)
+	tx.Add(&knftables.Rule{
+		Chain: kubeFirewallCheckChain,
+		Rule: knftables.Concat(
+			ipX, "daddr", ".", "meta l4proto", ".", "th dport", "@", kubeFirewallSet,
```

Comment:
maybe we can concatenate these strings 
```go
"daddr . meta l4proto . th dport @"
```
instead of
```go
"daddr", ".", "meta l4proto", ".", "th dport", "@"
```
As the buffer is not exposed by library, can't be reset and concatenating will reduce the number of mallocs ... 
https://github.com/danwinship/knftables/blob/fd49603c0a3d6012383657f44cae84781245a1a8/transaction.go#L57

Reply:
I went back and forth on this a little and eventually decided that I liked it better with the arguments split up semantically rather than doing as much "pre-concatenating" as I could.

> As the buffer is not exposed by library, can't be reset(reused) and concatenating will reduce the number of mallocs ...

Of course, we could decide to redo this later if that becomes an issue. See also [this comment](https://github.com/kubernetes/kubernetes/pull/121046#discussion_r1367736334) for other ideas.